HoRNDIS.cpp

```diff
@@ -28,516 +30,343 @@
+					confDesc, intDesc, candidate)) != NULL) {
+			const bool isEPIn =
+				(candidate->bEndpointAddress & kEndpointDescriptorDirection) != 0;
+			IOUSBHostPipe *&pipe = isEPIn ? fInPipe : fOutPipe;
+			if (pipe == NULL) {
+				// Note, 'copyPipe' already performs 'retain': must not call it again.
+				pipe = fDataInterface->copyPipe(candidate->bEndpointAddress);
+			}
+		}
+		if (fInPipe == NULL || fOutPipe == NULL) {
+			LOG(V_ERROR, "Could not init IN/OUT pipes in the Data Interface");
+			return false;
+		}
 	}
-	LOG(V_PTR, "PTR: fOutPipe: %p", fOutPipe);
-	LOG(V_DEBUG, "bulk output pipe %p: max packet size %d, interval %d", fOutPipe, epReq.maxPacketSize, epReq.interval);
 	
-	/* Currently, we don't even bother to listen on the interrupt pipe. */
-	
-	/* And we're done!  Wasn't that easy?*/
 	return true;
-	
-bailout5:
-	fDataInterface->release();
-bailout4:
-	fDataInterface->close(this);
-bailout3:
-	fDataInterface = NULL;
-bailout2:
-	fCommInterface->close(this);
-bailout1:
-	fCommInterface->release();
-	fCommInterface = NULL;
-bailout0:
-	/* Show what interfaces we saw. */
-	IOUSBFindInterfaceRequest req;
+}
 
-	req.bInterfaceClass = kIOUSBFindInterfaceDontCare;
-	req.bInterfaceSubClass = kIOUSBFindInterfaceDontCare;
-	req.bInterfaceProtocol = kIOUSBFindInterfaceDontCare;
-	req.bAlternateSetting  = kIOUSBFindInterfaceDontCare;
-	
-	IOUSBInterface *ifc = NULL;
-	LOG(V_ERROR, "openInterfaces: before I fail, here are all the interfaces that I saw, in case you care ...");
-	while ((ifc = fpDevice->FindNextInterface(ifc, &req))) {
-		LOG(V_ERROR, "openInterfaces:   class 0x%02x, subclass 0x%02x, protocol 0x%02x", ifc->GetInterfaceClass(), ifc->GetInterfaceSubClass(), ifc->GetInterfaceProtocol());
+void HoRNDIS::closeUSBInterfaces() {
+	fReadyToTransfer = false;  // Interfaces are about to be closed.
+	// Close the interfaces - this would abort the transfers (if present):
+	if (fDataInterface) {
+		fDataInterface->close(this);
 	}
-	
-	return false;
+	if (fCommInterface) {
+		fCommInterface->close(this);
+	}
+
+	OSSafeReleaseNULL(fInPipe);
+	OSSafeReleaseNULL(fOutPipe);
+	OSSafeReleaseNULL(fDataInterface);
+	OSSafeReleaseNULL(fCommInterface);  // First one to open, last one to die.
 }
 
 IOService *HoRNDIS::probe(IOService *provider, SInt32 *score) {
-	LOG(V_NOTE, "probe: came in with a score of %d", *score);
-	IOUSBDevice *dev = OSDynamicCast(IOUSBDevice, provider); /* XXX: error check */
-	
-	/* Do we even have one of the things that we can match on?  If not, return NULL.  Either 2/255/2, or 224/3/1.  */
-	const IOUSBConfigurationDescriptor *cd;
-	
-	cd = dev->GetFullConfigurationDescriptor(0);
-	if (!cd) {
-		LOG(V_ERROR, "probe: failed to get a configuration descriptor for configuration 0?");
+	LOG(V_DEBUG, "came in with a score of %d", *score);
+	
+	// Driver matching algorithm: keep it simple for now. Assumptions:
+	//  - Android device has only one USB configuration OR RNDIS interfaces
```

Comment:
I thought along the same lines. My reasons:
 - I am not sure that such devices exist AND would be used with this driver: not worth complicating the code.
 - I would not have any way to test that code => most likely it would be broken to begin with.

Reply:
The main non-phone device where we don't want it to regress is the BeagleBone series. 

I have several of those things here. What information from ioreg or other utilities would be helpful to determine what is needed to match to them?

Reply:
@d235j, thanks for the comment, the BeagleBone info would be helpful.

If you have a linux machine which you can connect the board to, please run "lsusb -v" on it and send me the output (you can limit it to the BeagleBoard if you know how).

If you only have Mac, then I could not find how to get the right information from system tools. You can try to get ported "lsusb" from either:
 - MacPorts, if you have them already, or don't mind installing them.
 - https://github.com/barisbalic/lsusb-osx - follow the instructions to build.
Also, if you want, I can send you the "lsusb" binary I've built on my machine, so you can use that.